test/e2e/framework/get.go

```diff
@@ -0,0 +1,131 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/onsi/gomega"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GetFunc is a function which retrieves a certain object.
+type GetFunc[T any] func(ctx context.Context) (T, error)
+
+// APIGetFunc is a get functions as used in client-go.
+type APIGetFunc[T any] func(ctx context.Context, name string, getOptions metav1.GetOptions) (T, error)
+
+// HandleRetry wraps an arbitrary get function. When the wrapped function
+// returns an error, HandleGetError will decide whether the call should be
+// retried and if requested, will sleep before doing so.
+//
+// This is meant to be used inside [gomega.Eventually] or [gomega.Consistently].
+func HandleRetry[T any](get GetFunc[T]) GetFunc[T] {
+	return func(ctx context.Context) (T, error) {
+		t, err := get(ctx)
+		if err != nil {
+			if retry, delay := ShouldRetry(err); retry {
+				if delay > 0 {
+					// We could return
+					// gomega.TryAgainAfter(delay) here,
+					// but then we need to funnel that
+					// error through any other
+					// wrappers. Waiting directly is simpler.
+					ctx, cancel := context.WithTimeout(ctx, delay)
+					defer cancel()
+					<-ctx.Done()
+				}
+				return t, err
+			}
+			// Give up polling immediately.
+			var null T
+			return t, gomega.StopTrying(fmt.Sprintf("Unexpected final error while getting %T", null)).Wrap(err)
+		}
+		return t, nil
+	}
+}
+
+// RetryNotFound wraps an arbitrary get function. When the wrapped function
+// encounters a "not found" error, that error is treated as a transient problem
+// and polling continues.
+//
+// This is meant to be used inside [gomega.Eventually] or [gomega.Consistently].
+func RetryNotFound[T any](get GetFunc[T]) GetFunc[T] {
+	return func(ctx context.Context) (T, error) {
+		t, err := get(ctx)
+		if apierrors.IsNotFound(err) {
+			// If we are wrapping HandleRetry, then the error will
+			// be gomega.StopTrying. We need to get rid of that,
+			// otherwise gomega.Eventually will stop.
+			var stopTryingErr gomega.PollingSignalError
+			if errors.As(err, &stopTryingErr) {
+				if wrappedErr := errors.Unwrap(stopTryingErr); wrappedErr != nil {
+					err = wrappedErr
+				}
+			}
+
+			// Mark the error as transient in case that we get
+			// wrapped by HandleRetry.
+			err = transientError{error: err}
+		}
+		return t, err
+	}
+}
+
+// GetObject takes a get function like clientset.CoreV1().Pods(ns).Get
```

Comment:
this is a personal stylistic nit that I have that you're welcome to ignore:
I personally prefer when functions are defined above functions they call--so GetObject would be defined above HandleRetry. This is so the file can be read top to bottom and you see the call stack. Up to you if you wanna adopt it

Reply:
I usually do that, too, but I am not strict about it. When I believe that a file is easier to understand when introducing some concept first (like handling retries), then I might introduce that first and then build on it with other functions.

However, in this case I've been inconsistent. Either GetObject should come at the top (as you said) or at the end, after all the retry functions. I'll probably move it to the top.